### PR TITLE
drivers: modem: HL7800: Allow operation without a SIM card

### PR DESCRIPTION
**Pull request by rerickson1**
Allow the driver to run if a SIM card is not present.
This allows public HL7800 APIs like firmware updates
to be used even if no network is available.
Remove duplicate query ICCID command.

### CODE REVIEW

File: drivers/modem/Kconfig.hl7800 [hunk 1]
```diff
@@ -265,10 +265,28 @@ config MODEM_HL7800_INIT_PRIORITY
...
 		When true the application must call mdm_hl7800_reset.
+		This mode delays the validity of modem info.
+
+config MODEM_HL7800_BOOT_IN_AIRPLANE_MODE
+	bool "Enter airplane mode during init"
+	help
+		mdm_hl7800_set_functionality must be used by app to exit
+		airplane mode.
+		If bands are reconfigured, then this setting is overridden
+		because modem is rebooted into full functionality mode.
```
Comment (mniestroj) [inline drivers/modem/Kconfig.hl7800:288]: Indentation issues.

File: drivers/modem/Kconfig.hl7800 [hunk 2]
```diff
@@ -288,8 +288,15 @@ config MODEM_HL7800_BOOT_IN_AIRPLANE_MODE
 
 endchoice
 
-config GET_IP_ADDR_INFO_ATTEMPTS
+config MODEM_HL7800_GET_IP_ADDR_INFO_ATTEMPTS
 	int "Number of attempts to get IP address info during init"
 	default 4
 
+config MODEM_HL7800_GPS
+	bool "Enable GPS command and handlers"
```
Comment (mniestroj) [inline drivers/modem/Kconfig.hl7800:297]: Maybe we should use GNSS instead of GPS as a more general term?

Adding GNSS should be really a separate PR. The thing is that I would much prefer to establish a generic API for GNSS providers, like proposed in #30564. However to achieve that, we first need to have list of initial GNSS modules we want to handle with proposed API to make sure we start with something scalable.

**Review by mniestroj — DISMISSED** [review]
First of all this PR needs to be split into several PRs, as there are many different unrelated changes here. Some are bugfixes, some are new features. Reviewing 16 commits is taking lots of time, so spliting into several smaller PRs will speed things up.

**Comment by rerickson1** [comment]
> First of all this PR needs to be split into several PRs, as there are many different unrelated changes here. Some are bugfixes, some are new features. Reviewing 16 commits is taking lots of time, so spliting into several smaller PRs will speed things up.

Thanks for the feedback. How will splitting into multiple PRs save time? There will still be that same amount of code to review and it will just create more work by dividing up the PR. The commits are small and targeted to their specific changes.  So reviewing each commit should still be possible.

**Comment by mniestroj** [comment]
>Thanks for the feedback. How will splitting into multiple PRs save time? There will still be that same amount of code to review and it will just create more work by dividing up the PR. The commits are small and targeted to their specific changes. So reviewing each commit should still be possible.

If you look from a developer point of view, then yes, it is simpler to push everything in one batch. In case of review process, there are much more code changes to look into. Big PRs are not encouraging potential reviewers, as it is lots of work in one go. Additionally even single comment on single commit can block (or just delay) whole PR, which in this case is a bunch of unrelated things (IMO). Then if there are multiple iterations, each time reviewers need to go through whole list of commits. Some reviewers might also be interested only in single functionality you add (e.g. GNSS) and are not likely to comment on other stuff. Splitting GNSS will definitely speed up bugfixes or small improvements from this PR, which do not need much effort to review and merge.

File: drivers/modem/hl7800.c [hunk 3]
```diff
@@ -4079,7 +4077,6 @@ static int modem_reset_and_configure(void)
 	LOG_ERR("Unable to configure modem");
 	ictx.configured = false;
 	set_network_state(HL7800_UNABLE_TO_CONFIGURE);
-	modem_reset();
```
Comment (rlubos) [inline drivers/modem/hl7800.c:4082]: Is this intentional to remove the reset on errors?

Comment (rerickson1) [reply drivers/modem/hl7800.c:4082]: Yes, we don't want to hold the modem in reset if initialization fails. This will allow an app to still communicate with the modem.

**Review by greg-leach — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
